File: swift_reference/Services/WeatherService.swift
import Foundation

actor WeatherService {
    static let shared = WeatherService()

    private let baseURL = "https://api.open-meteo.com/v1/forecast"
    private var cachedWeather: WeatherResponse?
    private var cachedLocation: (lat: Double, lon: Double)?
    private var cacheTimestamp: Date?
    private let cacheValidityMinutes: Int = 15

    private init() {}

    func fetchWeather(latitude: Double, longitude: Double, forceRefresh: Bool = false) async throws -> WeatherResponse {
        if !forceRefresh,
           let cached = cachedWeather,
           let cachedLoc = cachedLocation,
           let timestamp = cacheTimestamp,
           cachedLoc.lat == latitude,
           cachedLoc.lon == longitude,
           Date().timeIntervalSince(timestamp) < Double(cacheValidityMinutes * 60) {
            return cached
        }

        var components = URLComponents(string: baseURL)!
        components.queryItems = [
            URLQueryItem(name: "latitude", value: String(latitude)),
            URLQueryItem(name: "longitude", value: String(longitude)),
            URLQueryItem(name: "current", value: "temperature_2m,relative_humidity_2m,apparent_temperature,precipitation,weather_code,wind_speed_10m,wind_direction_10m,wind_gusts_10m"),
            URLQueryItem(name: "daily", value: "weather_code,temperature_2m_max,temperature_2m_min,precipitation_sum,precipitation_probability_max,wind_speed_10m_max,wind_gusts_10m_max"),
            URLQueryItem(name: "temperature_unit", value: "fahrenheit"),
            URLQueryItem(name: "wind_speed_unit", value: "mph"),
            URLQueryItem(name: "precipitation_unit", value: "inch"),
            URLQueryItem(name: "timezone", value: "auto"),
            URLQueryItem(name: "forecast_days", value: "7")
        ]

        guard let url = components.url else {
            throw WeatherError.invalidURL
        }

        let (data, response) = try await URLSession.shared.data(from: url)

        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            throw WeatherError.serverError
        }

        let decoder = JSONDecoder()
        let weather = try decoder.decode(WeatherResponse.self, from: data)

        cachedWeather = weather
        cachedLocation = (latitude, longitude)
        cacheTimestamp = Date()

        return weather
    }

    nonisolated func getDailyForecasts(from response: WeatherResponse) -> [DayForecast] {
        guard let daily = response.daily else { return [] }

        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd"

        var forecasts: [DayForecast] = []

        for i in 0..<min(daily.time.count, 7) {
            guard let date = dateFormatter.date(from: daily.time[i]) else { continue }

            let forecast = DayForecast(
                date: date,
                weatherCode: i < daily.weatherCode.count ? daily.weatherCode[i] : 0,
                highTemp: i < daily.temperature2mMax.count ? daily.temperature2mMax[i] : 0,
                lowTemp: i < daily.temperature2mMin.count ? daily.temperature2mMin[i] : 0,
                precipitationSum: i < daily.precipitationSum.count ? daily.precipitationSum[i] : 0,
                precipitationProbability: i < daily.precipitationProbabilityMax.count ? daily.precipitationProbabilityMax[i] : 0,
                maxWindSpeed: i < daily.windSpeed10mMax.count ? daily.windSpeed10mMax[i] : 0,
                maxWindGusts: i < daily.windGusts10mMax.count ? daily.windGusts10mMax[i] : 0
            )
            forecasts.append(forecast)
        }

        return forecasts
    }

    func calculateWindChill(temperature: Double, windSpeed: Double) -> Double {
        guard temperature <= 50 && windSpeed > 3 else {
            return temperature
        }

        let windChill = 35.74 + (0.6215 * temperature) - (35.75 * pow(windSpeed, 0.16)) + (0.4275 * temperature * pow(windSpeed, 0.16))
        return windChill
    }

    func clearCache() {
        cachedWeather = nil
        cachedLocation = nil
        cacheTimestamp = nil
    }
}

enum WeatherError: LocalizedError {
    case invalidURL
    case serverError
    case decodingError

    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid URL configuration"
        case .serverError:
            return "Weather server is unavailable"
        case .decodingError:
            return "Unable to parse weather data"
        }
    }
}

File: lib/services/weather_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/weather_data.dart';
import '../utils/constants.dart';

class WeatherService {
  WeatherResponse? _cachedWeather;
  double? _cachedLat;
  double? _cachedLon;
  DateTime? _cacheTimestamp;

  Future<WeatherResponse> fetchWeather({
    required double latitude,
    required double longitude,
    bool forceRefresh = false,
  }) async {
    // Check cache
    if (!forceRefresh &&
        _cachedWeather != null &&
        _cachedLat == latitude &&
        _cachedLon == longitude &&
        _cacheTimestamp != null &&
        DateTime.now().difference(_cacheTimestamp!).inMinutes <
            AppConstants.weatherCacheMinutes) {
      return _cachedWeather!;
    }

    final queryParams = {
      'latitude': latitude.toString(),
      'longitude': longitude.toString(),
      'current':
          'temperature_2m,relative_humidity_2m,apparent_temperature,precipitation,weather_code,wind_speed_10m,wind_direction_10m,wind_gusts_10m',
      'daily':
          'weather_code,temperature_2m_max,temperature_2m_min,precipitation_sum,precipitation_probability_max,wind_speed_10m_max,wind_gusts_10m_max',
      'temperature_unit': 'fahrenheit',
      'wind_speed_unit': 'mph',
      'precipitation_unit': 'inch',
      'timezone': 'auto',
      'forecast_days': '7',
    };

    final uri = Uri.parse(AppConstants.weatherApiBaseUrl)
        .replace(queryParameters: queryParams);

    try {
      final response = await http.get(uri);

      if (response.statusCode >= 200 && response.statusCode < 300) {
        final json = jsonDecode(response.body);
        final weather = WeatherResponse.fromJson(json);

        _cachedWeather = weather;
        _cachedLat = latitude;
        _cachedLon = longitude;
        _cacheTimestamp = DateTime.now();

        return weather;
      } else {
        throw WeatherException('Weather server error: ${response.statusCode}');
      }
    } catch (e) {
      if (e is WeatherException) rethrow;
      throw WeatherException('Unable to fetch weather: $e');
    }
  }

  List<DayForecast> getDailyForecasts(WeatherResponse response) {
    final daily = response.daily;
    if (daily == null) return [];

    final forecasts = <DayForecast>[];
    final count = daily.time.length.clamp(0, 7);

    for (int i = 0; i < count; i++) {
      final date = DateTime.tryParse(daily.time[i]);
      if (date == null) continue;

      forecasts.add(DayForecast(
        id: 'forecast_$i',
        date: date,
        weatherCode: i < daily.weatherCode.length ? daily.weatherCode[i] : 0,
        highTemp:
            i < daily.temperature2mMax.length ? daily.temperature2mMax[i] : 0,
        lowTemp:
            i < daily.temperature2mMin.length ? daily.temperature2mMin[i] : 0,
        precipitationSum:
            i < daily.precipitationSum.length ? daily.precipitationSum[i] : 0,
        precipitationProbability: i < daily.precipitationProbabilityMax.length
            ? daily.precipitationProbabilityMax[i]
            : 0,
        maxWindSpeed:
            i < daily.windSpeed10mMax.length ? daily.windSpeed10mMax[i] : 0,
        maxWindGusts:
            i < daily.windGusts10mMax.length ? daily.windGusts10mMax[i] : 0,
      ));
    }

    return forecasts;
  }

  double? calculateWindChill(double temperature, double windSpeed) {
    if (temperature > 50 || windSpeed <= 3) return null;

    return 35.74 +
        (0.6215 * temperature) -
        (35.75 * _pow(windSpeed, 0.16)) +
        (0.4275 * temperature * _pow(windSpeed, 0.16));
  }

  double _pow(double base, double exponent) {
    return base <= 0 ? 0 : _exp(exponent * _ln(base));
  }

  double _exp(double x) {
    double result = 1.0;
    double term = 1.0;
    for (int i = 1; i <= 100; i++) {
      term *= x / i;
      result += term;
      if (term.abs() < 1e-15) break;
    }
    return result;
  }

  double _ln(double x) {
    if (x <= 0) return double.negativeInfinity;
    double result = 0.0;
    while (x > 2) {
      x /= 2.718281828;
      result += 1;
    }
    while (x < 0.5) {
      x *= 2.718281828;
      result -= 1;
    }
    x -= 1;
    double term = x;
    for (int i = 1; i <= 100; i++) {
      result += (i % 2 == 1 ? 1 : -1) * term / i;
      term *= x;
      if (term.abs() < 1e-15) break;
    }
    return result;
  }

  String getWindDirection(int degrees) {
    const directions = ['N', 'NE', 'E', 'SE', 'S', 'SW', 'W', 'NW'];
    final index = ((degrees + 22.5) / 45).floor() % 8;
    return directions[index];
  }

  void clearCache() {
    _cachedWeather = null;
    _cachedLat = null;
    _cachedLon = null;
    _cacheTimestamp = null;
  }
}

class WeatherException implements Exception {
  final String message;
  WeatherException(this.message);

  @override
  String toString() => message;
}
